src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -683,15 +704,21 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
     expectCompletedMerge ln (SingleRun r) = do
       traceWith tr $ AtLevel ln $ TraceExpectCompletedMergeSingleRun (runNumber $ Run.runRunFsPaths r)
       pure r
-    expectCompletedMerge ln (MergingRun var) = do
-      withMVar var $ \case
-        CompletedMerge r -> do
+    expectCompletedMerge ln (MergingRun _ _ var) = do
+      modifyMVarMasked var $ \case
```

Comment:
The `TraceExpectCompletedMerge` would be an excellent place to report how many merge steps were actually done, and how many credits were actually supplied, vs the max credits theoretically needed for a merge on this level.

We should be able to make those trace results match up with the scheduled merging prototype.

Reply:
> Why does this need modifyMVar or modifyMVarMasked? Can't it be done simply with readMVar? It's not changing the contents of the MVar. The tracing can be done outside of the lock.

On the one hand, this is an artifact of refactoring this bit of code a few times. On the other hand, this bit of code will start changing the contents in #435. This can happen when a merge is *just about* finished, but it does not get enough credits to go over the next threshold before `expectCompletedMerge` is called. `expectCompletedMerge` will then have to the last bit of merging work. Nevertheless, I'll see if I can change this bit of code

Reply:
If we supply credits before flushing (which calls `expectCompletedMerge`), then I don't see how this should happen. Is it because there are `N` entries left to be merged, we supply `N` credits, but `N` is smaller than the threshold, so we don't do anything yet? In that case, it would be nice to finish the merge once we have enough credits, ignoring the threshold, but I suspect the reason against doing that is performance (having to look deeper into the structure to find out whether the credits complete the merge)? I feel like there should be a way to make this work without much overhead.

Reply:
I think it's fine if `expectCompletedMerge` does a bit of merging work if required. It's relatively simple, and only requires a little bit of extra code

Reply:
> Why does this need modifyMVar or modifyMVarMasked? Can't it be done simply with readMVar? It's not changing the contents of the MVar. The tracing can be done outside of the lock.

I've changed it to use `withMVar` for now, so that we put back the contents in the `error` case. I've also moved the tracing outside.

Reply:
Oh I see, yes, we will have supplied enough credits for the merge to complete, but we may not have gone over the threshold to spend those credits. So yes, I agree it's reasonably to do the remaining merging work in `expectCompletedMerge`, though for sanity checking we should spend just the remaining credits and then assert that the merge is complete (or equivalent assertion).